### PR TITLE
GAP-2105 | Change Question Type for Application Builder

### PR DESCRIPTION
**Pull request by ConorFayleTCO**
## Description

[GAP-2105](https://technologyprogramme.atlassian.net/browse/GAP-2105)

Summary of the changes and the related issue. List any dependencies that are required for this change:
https://github.com/cabinetoffice/gap-find-admin-backend/pull/199

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [x] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [x] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

[GAP-2105]: https://technologyprogramme.atlassian.net/browse/GAP-2105?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx [hunk 1]
```diff
@@ -61,36 +91,57 @@ export const getServerSideProps: GetServerSideProps = async ({
...
+            options,
+            validation: {
+              mandatory: optional !== 'true',
+              maxWords: '',
+            },
+          }
+        );
 
+        const editQueryString = query.backTo
+          ? `?${new URLSearchParams({ backTo: query.backTo.toString() })}`
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx:119]: template literal makes this a little harder to read

```suggestion
          ? '?' + new URLSearchParams({ backTo: query.backTo.toString() })
```

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.page.tsx [hunk 2]
```diff
@@ -32,40 +37,94 @@ export const getServerSideProps = async (
       (section) => section.sectionId === sectionId
     ) as ApplicationFormSection;
 
+    const queryString = context.query
+      ? `?${new URLSearchParams({ ...(context.query as object) })}`
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.page.tsx:41]: redundant object spread here

```suggestion
      ? `?${new URLSearchParams(context.query as object)}`
```

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/index.page.tsx [hunk 3]
```diff
@@ -14,13 +14,14 @@ const QuestionType = ({
...
 }: InferProps<typeof getServerSideProps>) => {
   return (
     <>
       <Meta
-        title={`${
-          fieldErrors.length > 0 ? 'Error: ' : ''
-        }Add a question - Manage a grant`}
+        title={`${fieldErrors.length > 0 ? 'Error: ' : ''}${
+          isEdit ? 'Change question type' : 'Add a question'
+        } - Manage a grant`}
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/index.page.tsx:24]: might be worth sticking this in a `getPageTitle` function

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx [hunk 4]
```diff
@@ -46,12 +57,31 @@ export const getServerSideProps: GetServerSideProps = async ({
...
+      sectionId,
+      questionId.toString()
+    );
+    if (questionData?.options) {
+      questionSummary = {
+        ...questionSummary,
+        options: questionData.options,
+        fieldTitle: questionData.fieldTitle,
+      };
+      options = questionSummary.options;
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx]: are we not always expecting to get `options` back here?

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/[questionId]/edit/question-options.getServerSideProps.tsx [hunk 5]
```diff
@@ -30,14 +32,34 @@ const getServerSideProps = async ({
     string,
     string
   >;
-  const { backTo } = query as Record<string, string>;
+  const { backTo, from } = query as Record<string, string>;
 
   function getBackToRedirect() {
+    if (from === 'question-type') {
+      const queryString = buildQueryStringWithoutUndefinedValues({ backTo });
+      console.log(queryString);
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/[questionId]/edit/question-options.getServerSideProps.tsx:40]: Can we remove this?

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/index.getServerSideProps.tsx [hunk 6]
```diff
@@ -46,78 +87,125 @@ function getRadio(responseType: ResponseType) {
...
+          },
+        }
+      );
+
+      return {
+        data: 'QUESTION_UPDATED',
+        sessionId,
+      };
+    } else {
+      if (redirectQuestionType.includes(body.responseType)) {
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/index.getServerSideProps.tsx:134]: We do this check on both sides of the if conditional, can we move it above the check for questionId?

**Review by dylanwrightCO — DISMISSED** [review]

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/[questionId]/edit/question-options.getServerSideProps.test.tsx [hunk 7]
```diff
@@ -186,6 +184,51 @@ describe('Question Options', () => {
...
+
+        expect(result.props.cancelChangesHref).toStrictEqual(
+          '/build-application/testApplicationId/dashboard'
+        );
+        expect(result.props.backButtonHref).toStrictEqual(
+          '/build-application/testApplicationId/testSectionId/testQuestionId/edit/question-content?backTo=dashboard'
+        );
+      });
+
+      it('Should return correct button links', async () => {
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/[questionId]/edit/question-options.getServerSideProps.test.tsx:208]:
```suggestion
      it('Should return correct button links when "from" passed in query', async () => {
```

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/[questionId]/edit/question-options.getServerSideProps.test.tsx [hunk 8]
```diff
@@ -186,6 +184,51 @@ describe('Question Options', () => {
...
+          })
+        )) as GetServerSideProps;
+
+        expect(result.props.cancelChangesHref).toStrictEqual('');
+        expect(result.props.backButtonHref).toStrictEqual(
+          '/build-application/testApplicationId/testSectionId/question-type?questionId=testQuestionId&sectionId=testSectionId'
+        );
+      });
+
+      it('Should return correct button links', async () => {
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/[questionId]/edit/question-options.getServerSideProps.test.tsx:221]:
```suggestion
      it('returns correct button links when "backTo" and "from" passed in query', async () => {
```

File: packages/admin/src/utils/QuestionPageGetServerSideProps.tsx [hunk 9]
```diff
@@ -16,22 +16,17 @@ import {
...
  *
- * @param context - GetServerSidePropsContext
- * @param fetchPageData - A function that takes in a jwt and returns data asynchronously
- * @param handleRequest - A function that takes in a jwt and the body the page returns, then updates/posts this data
- * @param jwt - A JWT needed for calls to the backend
- * @param onSuccessRedirectHref - Where to redirect to after successfully updating data
- * @param onErrorMessage - An error message to display if getting/updating data fails
  *
  * @returns A redirect to the relevant location, or a set of props needed to load a page
+ * @param props
```
Comment (jgunnCO) [inline packages/admin/src/utils/QuestionPageGetServerSideProps.tsx:21]: move this line above `@returns`?

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx [hunk 10]
```diff
@@ -66,7 +69,7 @@ describe('getServerSideProps()', () => {
     });
   });
 
-  it('should call postQuestion service with a POST request', async () => {
+  it.only('should call postQuestion service with a POST request', async () => {
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx:72]: rm `.only`?

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx [hunk 11]
```diff
@@ -87,6 +90,48 @@ describe('getServerSideProps()', () => {
     );
   });
 
+  it.only('should call patchQuestion service with a PATCH request', async () => {
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx:93]: rm `.only`?

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx (hunk 10 above)
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx:72]:
```suggestion
  it('should call postQuestion service with a POST request', async () => {
```

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx (hunk 11 above)
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-type/add-word-count.test.tsx:93]:
```suggestion
  it('should call patchQuestion service with a PATCH request', async () => {
```

**Review by jgunnCO — APPROVED** [review]

**Pull request merged**